### ISSUE TITLE
Weird behaviours of exit handlers in loop

### ISSUE THREAD

**Issue by laymain** [argoproj/argo-workflows#5463]
## Summary 

I'm trying to have a step that is executed at the end of each iteration of a loop whether or not it succeed or failed.
I'm struggling to add an exit handler for a loop iteration, every try I gave had a weird behaviour.

## Diagnostics

Argo version: v2.12.10

---

* First try, adding an exit handler on the iteration step

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: onexit-iteration-
spec:
  arguments:
    parameters:
    - name: dates
      value: |
        ["2021-03-01", "2021-03-02"]
  onExit: end-of-workflow
  entrypoint: onexit-loop
  templates:
  - name: onexit-loop
    parallelism: 1
    steps:
    - - name: iteration
        template: iteration
        onExit: end-of-iteration
        withParam: "{{workflow.parameters.dates}}"

  - name: iteration
    steps:
    - - name: step1
        template: succeed-step

  - name: succeed-step
    container:
      image: alpine
      command: ['/bin/sh', '-c']
      args: ["echo hello; exit 0"]

  - name: end-of-iteration
    container:
      image: docker/whalesay
      command: [ cowsay ]
      args: [ "iteration cleanup" ]

  - name: end-of-workflow
    container:
      image: docker/whalesay
      command: [ cowsay ]
      args: [ "workflow cleanup" ]
```

```d
STEP                                 TEMPLATE          PODNAME                            DURATION  MESSAGE
 ✔ onexit-iteration-bf6sb            onexit-loop
 ├─┬─✔ iteration(0:2021-03-01)       iteration
 │ │ └───✔ step1                     succeed-step      onexit-iteration-bf6sb-3620605730  6s
 │ └─✔ iteration(1:2021-03-02)       iteration
 │   └───✔ step1                     succeed-step      onexit-iteration-bf6sb-24746574    6s
 ├─✔ iteration(0:2021-03-01).onExit  end-of-iteration  onexit-iteration-bf6sb-341808317   8s
 └─✔ iteration(1:2021-03-02).onExit  end-of-iteration  onexit-iteration-bf6sb-348817417   8s

 ✔ onexit-iteration-bf6sb.onExit     end-of-workflow   onexit-iteration-bf6sb-667038399   7s
```

I was not expecting all exit handlers to be called successively **at the end** of the loop instead of after each iteration.

---

* Second try, adding an exit handler on a sub-iteration step

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: onexit-loop-
spec:
  arguments:
    parameters:
    - name: dates
      value: |
        ["2021-03-01", "2021-03-02", "2021-03-03"]
  onExit: end-of-workflow
  entrypoint: onexit-loop
  templates:
  - name: onexit-loop
    parallelism: 1
    steps:
    - - name: iteration
        template: iteration
        withParam: "{{workflow.parameters.dates}}"

  - name: iteration
    steps:
    - - name: iteration-steps
        template: iteration-steps
        onExit: end-of-iteration

  - name: iteration-steps
    steps:
    - - name: step1
        template: succeed-step

  - name: succeed-step
    container:
      image: alpine
      command: ['/bin/sh', '-c']
      args: ["echo hello; exit 0"]

  - name: end-of-iteration
    container:
      image: docker/whalesay
      command: [ cowsay ]
      args: [ "iteration cleanup" ]

  - name: end-of-workflow
    container:
      image: docker/whalesay
      command: [ cowsay ]
      args: [ "workflow cleanup" ]
```

```d
STEP                             TEMPLATE          PODNAME                       DURATION  MESSAGE
 ✔ onexit-loop-h6jlf             onexit-loop
 └─┬─✔ iteration(0:2021-03-01)   iteration
   │ ├───✔ iteration-steps       iteration-steps
   │ │   └───✔ step1             succeed-step      onexit-loop-h6jlf-559914117   6s
   │ └─✔ iteration-steps.onExit  end-of-iteration  onexit-loop-h6jlf-1231102337  7s
   ├─✔ iteration(1:2021-03-02)   iteration
   │ └───✔ iteration-steps       iteration-steps
   │     └───✔ step1             succeed-step      onexit-loop-h6jlf-2282645093  5s
   └─✔ iteration(2:2021-03-03)   iteration
     └───✔ iteration-steps       iteration-steps
         └───✔ step1             succeed-step      onexit-loop-h6jlf-2158012477  5s

 ✔ onexit-loop-h6jlf.onExit      end-of-workflow   onexit-loop-h6jlf-3488764721  7s
```

I was not expecting the exit handler to be called only once.

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by simster7** [comment]
Found the issue

**Comment by simster7** [comment]
Working on fix

**Comment by laymain** [comment]
That's great.

Btw, the fix seems to concern only the second case, is the first case result the expected one?

**Comment by simster7** [comment]
No, the first is not expected. The behavior you are seeing is slightly undefined and is a result of `parallelism: 1`. It's not clear what the expected behavior should be in that scenario. The main question is: if you set `parallelism: 1`, do you expect `onExit` nodes to also respect that? E.g. run `A -> A.onExit -> B`, or `A -> A.onExit, B`.

I may split the PR up into two so that we can fix the second case quickly while we investigate more of the first.

**Comment by laymain** [comment]
Perfect!

**Comment by laymain** [comment]
Any update on the second issue?

**Comment by alexec** [comment]
@laymain can you please open a new issue for your second issue?
